api/grpc/Server.go

```diff
@@ -5,6 +5,7 @@ import (
 	"authz/api"
 	core "authz/api/gen/v1alpha"
 	"authz/application"
+	"authz/domain/contracts"
```

Comment:
I think we are working against our agreed architecture in referring to "domain" objects in the api layer. We are already using the Principal model here, which in retrospect seems wrong. We should probably just use a string (or a type from *AppService instead), to avoid surfacing domain the objects here.

This PR adds contracts and a PrincipalRepo to the api layer where it seems like we should access it via a UserAppService in the app services layer instead.

Reply:
The rule, as I understand it generally, is that all direct references are to the same or a lower level and never to a higher level. So, taking that for granted, skipping a layer should be okay.

But two counterpoints to my own point:
1. I don't know that we should take that for granted. Maybe we want a stricter layering.
2. The operation the principalrepository is floating up here for probably doesn't really belong on it- it's resolving an access token (placeholder: actually just the username as a string) to an identity. In actual use, that's probably either doing some token introspection to extract an id or calling an introspection/userinfo endpoint on an idp, all of which sound like protocol-layer concerns.

I mainly did it this way because I did it this way two weeks ago when there wasn't as much separation, but I'm quickly talking myself into this being a protocol-layer utility of some kind. Maybe a UserAppService that takes a PrincipalRepository that gets implemented in infra and a TokenIdentityResolver or something that gets implemented in api so it can decode the token and look up the user without being dependent on the inner workings of either?

Reply:
Yeah, to be honest, I'm still absorbing our rules myself. :) One reason I don't like the idea of skipping a layer is because we had quite a detailed discussion at the F2F about the type of object we would send to the app services layer, e.g. CheckRequest, which takes the form of a non-protocol-specific request & dumb payload rather than a rich domain object. This feels right to me and should in theory reduce the burden of refactoring -- albeit at the cost of converting objects.

Maybe it's worth a quick discussion somewhere.